### ISSUE TITLE
Integration tests can not test assertion behaviors for all of NIO

### ISSUE THREAD

**Issue by gwynne** [apple/swift-nio#1509]
Pretty much what the title says. The integration tests can only test the crashing-ness (or not) of assertions in a very small subset of NIO's code (and of course the unit tests can't test it at all). This should be fixed and tests added for such methods as `EventLoop.preconditionInEventLoop()` and friends.

(Filed by request of @weissi and according to remarks in #1508)

**Comment by weissi** [comment]
Correct. When I first wrote [tests_05_assertions](https://github.com/apple/swift-nio/blob/master/IntegrationTests/tests_05_assertions) I wanted to save compile time in CI by constructing a reduced NIO package (`https://github.com/apple/swift-nio/blob/master/IntegrationTests/tests_05_assertions/defines.sh`).

That was a bad idea, we should just take the 30 seconds to compile NIO and build a package that can `import NIO` instead of trying to cherry-pick just the assertions.

This shouldn't be hard at all really, we need to build a SwiftPM package (much like we're doing now) but we should just add

```
dependencies: [
    .package("../../..")
]
```

or so to reach `swift-nio`. If that doesn't work then we can just use `swift package edit --path /path/to/NIO swift-nio` like we do in the allocation counter tests.
